### PR TITLE
✨ Add custom cookie expiration

### PR DESCRIPTION
**Pull request by jonathantyng-amp**
Add ability to set custom cookie expiration duration

See #22422 for more information

cc @zhouyx

### CODE REVIEW

File: extensions/amp-analytics/0.1/cookie-writer.js [hunk 1]
```diff
@@ -118,13 +118,24 @@ export class CookieWriter {
       return Promise.resolve();
     }
 
+    // convert cookieMaxAge (sec) to milliseconds
+    const cookieExpireDateMs = hasOwn(inputConfig, 'cookieMaxAge')
```
Comment (zhouyx) [inline extensions/amp-analytics/0.1/cookie-writer.js]: How do we want to handle invalid `cookieMaxAge` value? 
`cookieMaxAge = '1234'` or `cookieMaxAge = 'invalid'`

There's the `#isValidCookieConfig_` where I think is the right place to add assertion

File: extensions/amp-analytics/0.1/cookie-writer.js [hunk 2]
```diff
@@ -165,9 +176,10 @@ export class CookieWriter {
    * Expand the value and write to cookie if necessary
    * @param {string} cookieName
    * @param {string} cookieValue
+   * @param {number} cookieExpireDateMs
```
Comment (zhouyx) [inline extensions/amp-analytics/0.1/cookie-writer.js]: nit: looks like the value could be undefined. Maybe 
`@param {number=} opt_cookieExpireDateMs`?

File: extensions/amp-analytics/0.1/test/test-cookie-writer.js [hunk 3]
```diff
@@ -292,6 +292,28 @@ describes.fakeWin('amp-analytics.cookie-writer value', {amp: true}, env => {
     });
   });
 
+  it('should write cookie with custom expiration', () => {
```
Comment (zhouyx) [inline extensions/amp-analytics/0.1/test/test-cookie-writer.js]: Nice test! nit please also add test for invalid `cookieMaxAge` value. Thank you.

File: extensions/amp-analytics/0.1/cookie-writer.js (hunk 1 above)
Comment (jonathantyng-amp) [reply extensions/amp-analytics/0.1/cookie-writer.js]: I dont think `isValidCookieConfig_` is the best place for this, this function is used to check validity of cookie values. There's no overlap between that and validity of `cookieMaxAge`. I think adding a new method `getCookieMaxAge_(config)` would be cleaner.

Also, looking at how the rest of the code handles invalid numbers in the config, they do something like `Number(spec['totalTimeMin']) || 0`, trying to convert it to a number first then providing a default value if invalid.

Comment (zhouyx) [reply extensions/amp-analytics/0.1/cookie-writer.js]: I am seeing the `#isValidCookieConfig_` function as the place to check the validity of the entire configuration. But I think you make a good point on we don't want assertion, instead we want to fallback to a default value. 
`#getCookieMaxAge_()` works! Just one more request. If the value is not valid. Please report it as an user error, indicating that we expiration time falls back to default value due to invalid config. Thanks.

File: extensions/amp-analytics/0.1/cookie-writer.js [hunk 4]
```diff
@@ -149,10 +150,14 @@ export class CookieWriter {
       return BASE_CID_MAX_AGE_MILLIS;
     }
 
-    // convert cookieMaxAge (sec) to milliseconds
-    return (
-      Number(inputConfig['cookieMaxAge']) * 1000 || BASE_CID_MAX_AGE_MILLIS
-    );
+    const cookieMaxAgeNumber = Number(inputConfig['cookieMaxAge']);
+    if (!cookieMaxAgeNumber) {
+      user().error(TAG, 'invalid cookieMaxAge, falling back to default value');
```
Comment (zhouyx) [inline extensions/amp-analytics/0.1/cookie-writer.js]: nit: print out the invalid value as well and default value as well. 
`invalid cookieMaxAge %s, fallback to default value (1 year), inputConfig['cookieMaxAge']`

File: extensions/amp-analytics/0.1/cookie-writer.js [hunk 5]
```diff
@@ -149,10 +150,14 @@ export class CookieWriter {
       return BASE_CID_MAX_AGE_MILLIS;
     }
 
-    // convert cookieMaxAge (sec) to milliseconds
-    return (
-      Number(inputConfig['cookieMaxAge']) * 1000 || BASE_CID_MAX_AGE_MILLIS
-    );
+    const cookieMaxAgeNumber = Number(inputConfig['cookieMaxAge']);
+    if (!cookieMaxAgeNumber) {
```
Comment (zhouyx) [inline extensions/amp-analytics/0.1/cookie-writer.js]: nit: Few corner cases: 
cookeMaxAgeNumber = 123.12345, which 123123.45 couldn't convert to time. 
cookeMaxAgeNumber = 0
cookeMaxAgeNumber < 0

Could we add test cases accordingly. Thank you.

Comment (jonathantyng-amp) [reply extensions/amp-analytics/0.1/cookie-writer.js]: `new Date(123.123)` can handle decimals. Do we want to still consider it incorrect input?

File: extensions/amp-analytics/0.1/cookie-writer.js [hunk 6]
```diff
@@ -165,9 +198,10 @@ export class CookieWriter {
    * Expand the value and write to cookie if necessary
    * @param {string} cookieName
    * @param {string} cookieValue
+   * @param {number=} opt_cookieExpireDateMs
    * @return {!Promise}
    */
-  expandAndWrite_(cookieName, cookieValue) {
+  expandAndWrite_(cookieName, cookieValue, opt_cookieExpireDateMs) {
```
Comment (zhouyx) [inline extensions/amp-analytics/0.1/cookie-writer.js]: Looks to me that the input `cookieExpireDateMs` is always going to be a number. Do we still need to check for `opt_cookieExpireDateMs`?

File: extensions/amp-analytics/0.1/test/test-cookie-writer.js [hunk 7]
```diff
@@ -292,6 +292,139 @@ describes.fakeWin('amp-analytics.cookie-writer value', {amp: true}, env => {
...
+  });
+
+  it('should write cookie with default expiration (invalid string value)', () => {
+    win.location = 'https://www.example.com/';
+    const cookieWriter = new CookieWriter(
+      win,
+      win.document.body,
+      dict({
+        'cookies': {
+          'cookieMaxAge': 'invalid',
```
Comment (zhouyx) [inline extensions/amp-analytics/0.1/test/test-cookie-writer.js]: nit: May want to assert `allowConsoleError`or `expectAsyncConsoleError` functions to examine user error type.

**Comment by jonathantyng-amp** [comment]
cc @zhouyx ready to be merged

**Comment by jonathantyng-amp** [comment]
Discussed offline, we will allow cookieMaxAge <= 0

cc @zhouyx ready to be merged

**Comment by zhouyx** [comment]
LGTM! Merged.

**Comment by zhouyx** [comment]
Please also submit a PR to update doc : )
